### ISSUE TITLE
Define runtime config programmatically on server start

### ISSUE THREAD

**Issue by IlyaSemenov** [nuxt/nuxt.js#15366]

**Comment by chrisaddams** [comment]
I'm having a similar issue. I am running nuxt inside a docker container, running on kubernetes.
I set the environment variables of the container using a configmap, but I can't get nuxt to read these.

I have tried using @McPizzaIs plugin, which seems to work for setting static variables before start, but I can't seem to do this where I'm trying to read from something dynamic.

Using that plugin it should be something like this:

```export const result = (() => {

  const publicConfig = {
    myEnv: process.env.MY_ENV,
  };

  return { pass: true, publicConfig, continueOnFail: true };
})();
```
Though, I feel this should be something that works out of the box, right?
Shouldn't something like this work?
```
export default defineNuxtConfig({
    runtimeConfig: {
        myEnv: process.env.MY_ENV,
    }
});
```
And then started with `yarn start` inside of the docker container.

**Comment by McPizzaIs** [comment]
@chrisaddams My module is designed for loading config from external places, or validating config that exists.
It sound like you're just trying to load in config from docker.

If the `MY_ENV` is set in the docker container, it should work out of the box.
Have you prefixed `MY_ENV` with `NUXT_` (resulting in NUXT_MY_ENV) and added it to the `nuxt.config.ts` file as per https://nuxt.com/docs/getting-started/configuration

You can also ensure that the docker container has the right env variables by running:
`docker ps` then take note of the container ID running nuxt
then:
`docker exec CONTAINERID /usr/bin/env`

This will print out all the ENV vars that docker knows about to confirm its being set correctly

**Comment by chrisaddams** [comment]
@McPizzaIs Thanks for your help,
using docker exec the variables are definitely set.
I can't see anywhere in the documentation that it says to prefix the env variable with `NUXT_`?

**Comment by McPizzaIs** [comment]
Hmmm... How odd!
It was there as a requirement before
And the code still shows that nitro looks for a NUXT_

maybe something changed somewhere.
Have you tried logging out process.env?

**Comment by chrisaddams** [comment]
So this is definitely not working.
If the app is built with docker and started with `nuxt start` then the app will not pickup the environment variables from the running container. It will only take defaults from a baked in .env file. You cannot modify the .env file and have those picked up by the containerised app either.

**Comment by danielroe** [comment]
@chrisaddams If you are encountering this now working, please raise a new issue with a reproduction. (This issue is for tracking a different enhancement instead.) Note that Nuxt _will not_ read your `.env` file in production but will rely on you ensuring they are already defined in the environment, with the most appropriate method for your environment.

**Comment by chrisaddams** [comment]
Thanks @danielroe - I will raise an issue

**Comment by IlyaSemenov** [comment]
@McPizza0 From what I see, both your module and the https://github.com/johannschopplich/nuxt-prepare which you marked as a successor allow to inject code to the **build phase**.

What this ticket is about is the option to somehow inject custom code to the app **start phase** (i.e. during `node node-server.mjs`).

**Comment by maestroosram** [comment]
@IlyaSemenov did you find a workaround in the end? I am struggling with the same problem - and no one seem to understand it.

**Comment by aborn** [comment]
https://stackoverflow.com/questions/76488291/how-to-fetch-data-as-part-of-server-start-up-in-nuxt-3

**Comment by IlyaSemenov** [comment]
Basically what they suggest is to ditch Nuxt runtime config and use custom config object prepared in a Nitro plugin.

This approach doesn't solve the issue fully:
- It doesn't work for third party libs which continue to rely on Nuxt runtime config.
- It doesn't address the _"I would like to be able to catch that on server start and prevent further client-side runtime crashes"_ requirement.
- It doesn't work for server-side code which currently allows `useRuntimeConfig()` on module level.

**Comment by manniL** [comment]
(Linking https://github.com/nuxt/nuxt/pull/27117/ here)

**Comment by ellermister** [comment]
@danielroe 

I also hope to add support for this.

I hope to use the same docker image and set different environment variables to achieve different configurations during deployment.

for rc
```
version: '3'
services:
  web:
    image: myapp/nuxt
    environment:
      API_URL: http://rc.test.com
      API_KEY:  rc_key_***
```
for prod
```
version: '3'
services:
  web:
    image: myapp/nuxt
    environment:
      API_URL: http://prd.test.com
      API_KEY:  prd_key_***
```

**Comment by IlyaSemenov** [comment]
@ellermister What you want is already possible (and has been possible for years), just name your variables as `NUXT_API_URL` (not `API_URL`). See [Nuxt documentation](https://nuxt.com/docs/guide/going-further/runtime-config#example) for details.

This issue is about being able to additionally customize runtime config, by using _something else_ other than raw environment variables.

**Comment by IlyaSemenov** [comment]
I've just stumbled upon an another legitimate scenario for the proposed improve. I'm using two third party Nuxt modules which both need a server-side secret key (a random string) for their operation. They take it from runtime variables `NUXT_JWT_SECRET` and `NUXT_SESSION_PASSWORD`, respectively, and now I need to manage both.

I'd like to be able to only keep 1 respective env variable setup and then pass it to the two runtime config variables by hand.

**Comment by luc122c** [comment]
@IlyaSemenov Have you tried altering the cofniguration during a hook? The Nuxt SEO module is a good example of this type of config, [see this example](https://nuxtseo.com/docs/site-config/api/nuxt-hooks). If you can find the right hook, you should be alter the config before your middleware.

See:
- https://nuxt.com/docs/guide/going-further/hooks
- https://nuxt.com/docs/api/advanced/hooks

**Comment by IlyaSemenov** [comment]
@luc122c To my knowledge, there is no hook that would allow to alter runtime configuration. The example you've linked to clearly says that it will modify the **build time** site config. This issue is about altering config in run time (on app start) not on app build.

I've just checked again and in latest Nuxt 3.13.2, the following is literally imprinted to `.output/server/chunks/runtime.mjs` (parts removed for brevity):

```js
const _sharedRuntimeConfig = _deepFreeze(
  applyEnv(klona(_inlineRuntimeConfig), envOptions)
);

function useRuntimeConfig(event) {
  if (!event) {
    return _sharedRuntimeConfig;
  }
  // ...
}

const _inlineRuntimeConfig = {
  "app": {
    "baseURL": "/",
    // ...
```

So what it does is that it takes the literal value of the original runtime config template from `nuxt.config.ts`, calls `applyEnv` on its clone and then immediately freezes it. There is no hook called in between. `applyEnv` doesn't call any hooks either. The frozen config could not be altered by any code that goes after it.

**Comment by IlyaSemenov** [comment]
Of course not. For starters, `app.config.ts` is exposed to the client, so it's not suitable for holding app runtime config which includes secrets.

The use case is about **modifying runtime config** in run time (on app start), for the following purposes:
- Validation (prevent app from starting with invalid config).
- Normalization (e.g. ensure trailing slashes or replace empty strings with undefined).
- Fill from sources other than environment variables, such as AWS Secrets Manager.
- Custom logic for parsing environment variables, e.g. allow `NUXT_VAR=123` to be injected as string, or use non-`NUXT_` vars, or use single var for two runtime config settings.

Neither of the above is what `app.config.ts` is about.

**Comment by kaneda-1954** [comment]
[updateRuntimeConfig](https://nuxt.com/docs/api/kit/runtime-config#updateruntimeconfig) not enough?

**Comment by IlyaSemenov** [comment]
No. Again, `updateRuntimeConfig` is a **build-time** operation. It's a part of `@nuxt/kit` which is a set of build-time module helpers. It doesn't work in runtime. Even simply importing `@nuxt/kit` in runtime doesn't work:

```
RollupError: [plugin impound] This module cannot be imported in server runtime. [importing @nuxt/kit from server/plugins/worker.ts] 
```

As I elaborated above, it is not currently possible — even in theory — to modify the runtime config in, erhm, run-time. The runtime config is [prepared in a single shot and it's then immediately frozen](https://github.com/nitrojs/nitro/blob/1cc357fd76dc538a1dab468a13daaf7ab447c427/src/runtime/internal/config.ts#L20-L29). If there were some kind of a hook (ideally, an async hook) somewhere in between `applyEnv` and `deepFreeze`, that would suffice.

---

You do understand difference between runtime and build time, correct? Build time code runs when you do `nuxt build` and runtime code runs when you do `nuxt preview` (or `node .output/server/index.mjs`). This issue is about the latter, not the former.

**Comment by adrianrudnik** [comment]
Similar issue, would like to retrieve a token from a remote API on startup and store it within the runtime config, but well, its just not the right place.

While digging through node server mode with nitro, I found that you can abuse nitro plugins a bit, though not completely:

```ts
# server/plugins/modify-env.ts
export default defineNitroPlugin((nitroApp) => {
  process.env.NUXT_PUBLIC_YOUR_VAR = 'whatever'
})
```

is executed while the composable `useRuntime` is uninitialised, so the overridden ENV right here is actually taken into account in my `npm run preview`, and a later call to `useRuntime` will reflect this new value.

Not perfect, not stable enough. Maybe one could hack into the [auto import feature](https://github.com/unjs/unimport?tab=readme-ov-file#programmatic-usage) during runtime to return a facade composable instead or something.

**Comment by septatrix** [comment]
Our use case is reading secrets from files which is how systemd credentials and docker/kubernetes secrets are readable.

Environment variables are not sufficient for our use case (nor allowed in our deployment) because there are several security risks associated with them (inherited by child, picked up by logging e.g. journald, cannot be cleanly unset)

**Comment by danielroe** [comment]
this is a feature request probably for nitro - in fact there may be an issue tracking it already.

@pi0 already has some ideas for changes to runtime config in nitro v3 - tagging him in case he wants to add any more

**Comment by septatrix** [comment]
> I would personally use standard environment variables (`process.env`) for this case. And plan is to improve on this area (typing, validation).

So there are no plans to accommodate systemd credentials or Docker/K8s secrets? I think some hook to extend a dictionary would be very useful and also make reading secrets from files possible.

Bonus points if it would also be possible to trigger this during normal server runtime. This would be very useful for rotating tokens without shutting the server down

**Comment by ron-tayler** [comment]
I solved this problem for myself.
Example my runtimeConfig:
```
runtimeConfig: {
    public: {
      baseUrl: process.env.NUXT_PUBLIC_API_BASE ,
      institutionID: process.env.NUXT_PUBLIC_API_INST_ID,
      googleClientID: process.env.NUXT_PUBLIC_API_CLIENT_ID,
      foo: {
        barBaz: "qwerty"
      }
    },
    secretToken: ""
  },
```
This example using local .env file, but `NUXT_PUBLIC_API_YANDEX_CLIENT_ID` don't use in runtime docker container.

In runtime, Nitro will load the env. Nitro uses a configuration key tree, and add prefix `NITRO_`, but don't uses prefix `NUXT_`

if we want override institutionID in docker container:
Key in runtimeConfig: `public.institutionID`
Key in env: `NITRO_PUBLIC_INSTITUTION_ID`

if we want override googleClientID in docker container:
Key in runtimeConfig: `public.googleClientID`
Key in env: `NITRO_PUBLIC_GOOGLE_CLIENT_ID`

if we want override barBaz in docker container:
Key in runtimeConfig: `public.foo.barBaz`
Key in env: `NITRO_PUBLIC_FOO_BAR_BAZ`

if we want override secretToken in docker container:
Key in runtimeConfig: `secretToken`
Key in env: `NITRO_SECRET_TOKEN`

**Comment by IlyaSemenov** [comment]
> I solved this problem for myself.

I think you've misunderstood the problem and solved something else. I don't see how the approach you've described can solve any of the concerns raised in this issue: config values normalization, validation, populating from a custom external source.

**Comment by ron-tayler** [comment]
> I think you've misunderstood the problem and solved something else

You're right. I misunderstood the problem. And I solved my other problem, which is only indirectly related to the topic of this discussion.
